src/WidgetBase.ts

```diff
@@ -335,20 +335,24 @@ export class WidgetBase<P extends WidgetProperties> extends Evented implements W
 	 */
 	protected addDecorator(decoratorKey: string, value: any): void {
 		value = Array.isArray(value) ? value : [ value ];
-
-		let decoratorList = decoratorMap.get(this.constructor);
-		if (!decoratorList) {
-			decoratorList = new Map<string, any[]>();
-			decoratorMap.set(this.constructor, decoratorList);
+		if (this._properties) {
+			const decorators = this._decoratorCache.get(decoratorKey) || [];
+			this._decoratorCache.set(decoratorKey, [ ...decorators, ...value ]);
```

Comment:
You cannot use decorators and the programmatic API together, it's one or the other. In the get call it just returns the entry from `_decoratorCache` if it exists so adding them directly into there when dealing with an instance should be fine.

There is a sentence in the README indicating that the two methods should not be mixed.

I could however move the population of the `_decoratorCache` to the WidgetBase constructor now that using the API adds them directly into the instance's cache...